src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp

```diff
@@ -0,0 +1,795 @@
+/*
+* Crystals Dilithium Digital Signature Algorithms
+* Based on the public domain reference implementation by the
+* designers (https://github.com/pq-crystals/dilithium)
+*
+* Further changes
+* (C) 2021-2022 Jack Lloyd
+* (C) 2021-2022 Manuel Glaser, Michael Boric, René Meusel - Rohde & Schwarz Cybersecurity
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/dilithium.h>
+
+#include <botan/ber_dec.h>
+#include <botan/der_enc.h>
+#include <botan/hash.h>
+#include <botan/rng.h>
+#include <botan/exceptn.h>
+
+#include <botan/internal/dilithium_polynomials.h>
+#include <botan/internal/shake.h>
+#include <botan/internal/pk_ops_impl.h>
+#include <botan/internal/stl_util.h>
+#include <botan/internal/parsing.h>
+
+#include <algorithm>
+#include <iterator>
+#include <array>
+#include <vector>
+
+namespace Botan {
+namespace {
+
+/**
+ * Helper class to ease unmarshalling of concatenated fixed-length values
+ */
+template <typename AllocT>
+class BufferSlicer final
+   {
```

Comment:
I've used this to play a bit with `std::span`. I particularly like that this can make working with `std::array<>` much easier. I.e. having an API that takes `std::span<const uint8_t>` does not care whether the caller holds the data as a `std::vector<>` or a `std::array<>`.

But to fully reap the benefits of this, lots of (internal) APIs will need to be adapted to `std::span`.

Reply:
It seems to me totally reasonable to put this into `mem_ops.h` as a general utility. We can expand/modify it as necessary. Slicing buffers in this way is a very common operation.